modules/r2b2AnalyticsAdapter.js

```diff
@@ -0,0 +1,574 @@
+import {ajax} from '../src/ajax.js';
+import adapter from '../libraries/analyticsAdapter/AnalyticsAdapter.js';
+import {EVENTS} from '../src/constants.js';
+import adapterManager from '../src/adapterManager.js';
+import {getGlobal} from '../src/prebidGlobal.js';
+import {logWarn, logError, isNumber, isStr, isPlainObject} from '../src/utils.js';
+import {getRefererInfo} from '../src/refererDetection.js';
+import {config} from '../src/config.js';
+
+const ADAPTER_VERSION = '1.1.0';
+const ADAPTER_CODE = 'r2b2';
+const MODULE_NAME = 'R2B2 Analytics'
+const GVLID = 1235;
+const analyticsType = 'endpoint';
+
+const DEFAULT_SERVER = 'log.r2b2.cz';
+const DEFAULT_EVENT_PATH = 'prebid/events';
+const DEFAULT_ERROR_PATH = 'error';
+const DEFAULT_PROTOCOL = 'https';
+
+const ERROR_MAX = 10;
+const BATCH_SIZE = 50;
+const BATCH_DELAY = 500;
+const REPORTED_URL = getRefererInfo().page || getRefererInfo().topmostLocation;
+
+const START_TIME = Date.now();
+
+const EVENT_MAP = {};
+EVENT_MAP[EVENTS.NO_BID] = 'noBid';
+EVENT_MAP[EVENTS.AUCTION_INIT] = 'init';
+EVENT_MAP[EVENTS.BID_REQUESTED] = 'request';
+EVENT_MAP[EVENTS.BID_TIMEOUT] = 'timeout';
+EVENT_MAP[EVENTS.BID_RESPONSE] = 'response';
+EVENT_MAP[EVENTS.BID_REJECTED] = 'reject';
+EVENT_MAP[EVENTS.BIDDER_ERROR] = 'bidError';
+EVENT_MAP[EVENTS.BIDDER_DONE] = 'bidderDone';
+EVENT_MAP[EVENTS.AUCTION_END] = 'auction';
+EVENT_MAP[EVENTS.BID_WON] = 'bidWon';
+EVENT_MAP[EVENTS.SET_TARGETING] = 'targeting';
+EVENT_MAP[EVENTS.STALE_RENDER] = 'staleRender';
+EVENT_MAP[EVENTS.AD_RENDER_SUCCEEDED] = 'render';
+EVENT_MAP[EVENTS.AD_RENDER_FAILED] = 'renderFail';
+EVENT_MAP[EVENTS.BID_VIEWABLE] = 'view';
+
+/* CONFIGURATION */
+let WEBSITE = 0;
+let CONFIG_ID = 0;
+let CONFIG_VERSION = 0;
+let LOG_SERVER = DEFAULT_SERVER;
+
+/* CACHED DATA */
+let orderedAuctions = [];
+let auctionsData = {};
+let bidsData = {};
+let adServerCurrency = '';
+
+let flushTimer;
+let eventBuffer = [];
+let errors = 0;
+function flushEvents () {
+  let events = { prebid: { e: eventBuffer, c: adServerCurrency } };
+  eventBuffer = [];
+  reportEvents(events)
+}
+
+export function resetAnalyticAdapter() {
+  orderedAuctions = [];
+  auctionsData = {};
+  bidsData = {};
+  adServerCurrency = '';
+  clearTimeout(flushTimer);
+  eventBuffer = [];
+  errors = [];
+
+  WEBSITE = 0;
+  CONFIG_ID = 0;
+  CONFIG_VERSION = 0;
+  LOG_SERVER = DEFAULT_SERVER;
+}
+function processEvent (event) {
+  // console.log('process event:', event);
+  // console.log(JSON.stringify(event));
+  if (!event) {
+    return
+  }
+  eventBuffer.push(event);
+  if (flushTimer) {
+    clearTimeout(flushTimer);
+    flushTimer = null
+  }
+  if (eventBuffer.length >= BATCH_SIZE) {
+    flushEvents();
+  } else {
+    flushTimer = setTimeout(flushEvents, BATCH_DELAY);
```

Comment:
I think it will be highly likely the `processEvent` will be run many time concurrently given the number of events your adapter is registering to listen to and how some publishers can setup 1 auction per ad unit. have you stress tested this function in those conditions, it appears likely that `flushEvents` could be run multiple times simultaneously without a safeguard to prevent execution if already run. I'm not sure if the `BATCH_SIZE` and `BATCH_DELAY` will handle all cases.